### ISSUE TITLE
add_child / add_root not race condition safe

### ISSUE THREAD

**Issue by AnnaDamm** [django-treebeard/django-treebeard#53]
When calling add_child in multiple threads at the same time, a race condition occurs (on databases that check the unique key).
The path is generated in python code, so when a second ``add_child`` call is in progress just when a path for a different one is already generated, the second try to write into the database fails with a Duplicate Unique Key Error.
There should be some kind of database-lock be used to prevent this.

**Comment by zbyte64** [comment]
Is it possible to trigger this failure without threads merely by instantiating and saving objects in a different order? Example: We instantiate nodes [1,2,3,4,5] but they are not saved in the same order, maybe as [3,1,2,4,5].

I am having unique key errors even after wiping when running an import script. Pretty sure I should just use the bulk import instead of manually creating each instance....

**Comment by tolsac** [comment]
I have exactly the same problem; please do something about it

**Comment by onekiloparsec** [comment]
As far as I can tell, this also occurs on `.move`s.

**Comment by jacobbridges** [comment]
The problem doesn't seem to be limited to a multi-threaded environment. I'm attempting to load a large, nested comment structure into django-treebeard and running into the error:

```
IntegrityError: (1062, "Duplicate entry '00010002000B0002' for key 'path'")
```

Using MySQL 5.7. When adding children, I'm fetching the parent node each time (to avoid the [`Raw Query`](https://django-treebeard.readthedocs.io/en/latest/caveats.html#known-caveats) known caveat in the documentation).

**Comment by onekiloparsec** [comment]
I personally solved my problem with the pessimistic lock approach described here: https://medium.com/@hakibenita/how-to-manage-concurrency-in-django-models-b240fed4ee2

**Comment by stuaxo** [comment]
Hitting this on postgres with treebeard (via djangocms).

I have a data importer running via celery, each process deletes and then creates items, and I just get loads of duplicate entry errors.

**Comment by tabo** [comment]
I can see this happening in periods of heavy writes. A quick solution would be for the user to lock the row (or table) in their transaction. Locking the row is something we could consider in the library itself too, and calculating the new path after the lock is grabbed.

**Comment by stuaxo** [comment]
Please consider doing it in the library.   Libraries should be safe by default - if it needs a way for the user to disable that safety then fair enough.
Any unsafe operations should be documented, having stuff like this break breaks the abstraction.

**Comment by kursat-ozer** [comment]
We are having the same issue. When a user selects multiple folders to move, a request is made for each folder. This creates a race condition where threads are reading and calculating the new path. If two threads reads at the same time, they generates same path and the later one raises integrity errors on DB. 

`select_for_update` is not suitable for this case, because it doesn't lock the rows to reads. For each move, we need to wait for the previous move to complete (reads & writes). 

Locking table to reads also caused some other problems. Is there anyone solved the issue ?

**Comment by waveywhite** [comment]
This approach worked for me when inserting lots of new nodes caused the key error race condition:

```python
from treebeard.mp_tree import MP_Node, MP_AddChildHandler, MP_AddSiblingHandler, MP_MoveHandler

class MyTreeNode(MP_Node):
    # related fields here ...

    # override tree handling functions
    def add_child(self, task=None):
        with transaction.atomic():
            me = GanttTaskTreeNode.objects.select_for_update().get(id=self.id)
            return MP_AddChildHandler(me, task=task).process()

    def add_sibling(self, pos=None, task=None):
        with transaction.atomic():
            me = GanttTaskTreeNode.objects.select_for_update().get(id=self.id)
            return MP_AddSiblingHandler(me, pos=pos, task=task).process()
    
    def move(self, target, pos=None):
        with transaction.atomic():
            me = GanttTaskTreeNode.objects.select_for_update().get(id=self.id)
            return MP_MoveHandler(me, target=target, pos=pos).process()

```

**Comment by stuaxo** [comment]
Thanks @solarissmoke for getting this over the line, and everyone on here who contributed code + info, its good seeing this stuff improve.

**Comment by pcraciunoiu** [comment]
I tested the v5 release, and the race condition is still there for me for `add_child`.

The code in [the latest comment](https://github.com/django-treebeard/django-treebeard/issues/53#issuecomment-3670510191) doesn't raise any errors, but v5.0.1 still throws `IntegrityError` when adding multiple children to the same node in parallel.

I don't think this issue should be closed just yet.

**Comment by solarissmoke** [comment]
Hey @pcraciunoiu thanks for your feedback. I tried to address the scope of the changes in the [commit message](https://github.com/django-treebeard/django-treebeard/commit/5121b788ea0c9f9224cd12ddc1c0d6bf597d68da) that added atomic transaction wrappers:

> Transactions alone do not necessarily avoid these issues: for some
operations (e.g., moves), a lock would need to be placed on the entire
table to eliminate any risk. It is suggested that projects that need to
do so add such locking themselves, as there are performance
implications.

and also in the [documentation](https://django-treebeard.readthedocs.io/en/latest/caveats.html#inconsistent-state).

I stopped short of applying a whole-table lock, which could have potentially significant performance implications depending on use case. The sample code provided above was only doing a whole-table lock for adding root nodes however, so it is possible that there is some other issue with `add_child`. On that basis, if you are able to reproduce consistently and have time to contribute a PR that resolves it without requiring a table lock, then that would be appreciated!

**Comment by stuaxo** [comment]
> If a whole-table lock is the only way to resolve, I would not be opposed to adding support for that behind a setting (disabled by default), so that projects can enable it with an understanding of the performance implications.

For something like data ingest, it would be good to have this as an option.

**Comment by solarissmoke** [comment]
With thanks to @pcraciunoiu the race condition for `add_child` has now been fixed. 

I believe I also [found a way](https://github.com/django-treebeard/django-treebeard/pull/370) to make `add_root` safe without having to lock the whole table (just locking all root nodes seems to be sufficient, at least for the simple case of adding multiple roots concurrently). These changes will be shipped in v5.0.3. If anyone still runs into problems, please open a new issue.

**Comment by pcraciunoiu** [comment]
Thanks @solarissmoke, I tested the latest changes several times, and couldn't make it fail adding children. I even bumped concurrency to 8 celery workers and tried 300+ files repeatedly, and no IntegrityErrors were raised 🎉

**Comment by waveywhite** [comment]
Thank you @solarissmoke. These changes have solved the issues I was seeing.
